zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
     end
 end
-export strerror;
 
-# TODO: get this into docs:
-# Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, handler, usr)
-
-    handler_jl = __zcm_handler(handler, usr);
-    handler_c  = cfunction(__zcm_handler_wrapper, Void,
-                           (Ptr{Native.RecvBuf}, Cstring, Ptr{Void},));
-    uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"), Ptr{Native.UvSub},
-                       (Ptr{Void}, Any,), handler_c, handler_jl);
-    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"));
-
-    return Sub(ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
-                     (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub},),
-                     zcm.zcm, channel, uv_handler, uv_wrapper),
-               handler_jl, handler_c, uv_handler, uv_wrapper);
+function onresponse(rbuf::Native.RecvBuf, channelbytes::Cstring, opts::SubscriptionOptions)
+    channel = unsafe_string(channelbytes)
+    msgdata = unsafe_wrap(Vector{UInt8}, rbuf.data, rbuf.data_size)
+    msg = decode(opts.msgtype, msgdata)
+    opts.handler(channel, msg)
+    return nothing
+end
+
+function subscribe(zcm::Zcm, channel::AbstractString, options::T) where T <: SubscriptionOptions
+    # The C function is called with the receive buffer, channel name, and the 
+    # SubscriptionOptions struct
+    c_handler = cfunction(onresponse, Void, Tuple{Ref{Native.RecvBuf}, Cstring, Ref{T}})
+    uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"),
+                       Ptr{Native.UvSub}, 
+                       (Ptr{Void}, Ptr{Void}),
+                       c_handler, Ref(options))
+    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"))
+    csub = ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
+                (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub}),
+                zcm, channel, uv_handler, uv_wrapper)
+    sub = Subscription(options, c_handler, uv_wrapper, uv_handler, csub)
+    push!(zcm.subscriptions, sub)
+    return sub
 end
-export subscribe;
 
-# RRR: might be a way to make this template a bit prettier
 # TODO: get this into docs:
 # Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, msg::MsgType, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, MsgType::DataType, handler, usr)
-    return subscribe(zcm, channel,
-                     function (rbuf::RecvBuf, channel::AbstractString, usr)
-                         msg = decode(MsgType, rbuf.data)
-                         handler(rbuf, channel, msg, usr);
-                     end, usr);
+#               handler(channel::String, msg::msgtype) defined
+function subscribe(zcm::Zcm, channel::AbstractString, msgtype::Type{<:AbstractZCMType}, handler::Function)
```

Comment:
Again, why not allow callable types for `handler`?

Reply:
fixed, thanks